chain-signatures/node/src/indexer_eth/mod.rs

```diff
@@ -0,0 +1,1514 @@
+        }
+    }
+
+    async fn get_latest_block_number(&self) -> anyhow::Result<u64> {
+        match self {
+            EthereumClient::Helios(client) => client.get_latest_block_number().await,
+            EthereumClient::DirectRpc(client) => client.get_latest_block_number().await,
+        }
+    }
+
+    async fn get_transaction_receipt(
+        &self,
+        tx_hash: alloy::primitives::B256,
+    ) -> anyhow::Result<Option<alloy::rpc::types::TransactionReceipt>> {
+        match self {
+            EthereumClient::Helios(client) => client.get_transaction_receipt(tx_hash).await,
+            EthereumClient::DirectRpc(client) => client.get_transaction_receipt(tx_hash).await,
+        }
+    }
+}
+
+impl EthereumClient {
+    pub async fn new(eth: EthConfig) -> anyhow::Result<EthereumClient> {
+        if eth.light_client {
+            Ok(EthereumClient::Helios(
+                indexer_eth_helios::build_client(eth.clone()).await?,
+            ))
+        } else {
+            Ok(EthereumClient::DirectRpc(
+                indexer_eth_direct_rpc::RpcEthereumClient::new(&eth.execution_rpc_http_url),
+            ))
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct EthereumIndexer {
+    eth: EthConfig,
+    sign_tx: mpsc::Sender<Sign>,
+    app_data_storage: AppDataStorage,
+    node_near_account_id: AccountId,
+    backlog: Backlog,
+    contract_watcher: ContractStateWatcher,
+    mesh_state: watch::Receiver<MeshState>,
+    node_client: NodeClient,
+    client: EthereumClient,
+}
+
+impl EthereumIndexer {
+    #[allow(clippy::too_many_arguments)]
+    pub async fn new(
+        eth: Option<EthConfig>,
+        sign_tx: mpsc::Sender<Sign>,
+        app_data_storage: AppDataStorage,
+        node_near_account_id: AccountId,
+        backlog: Backlog,
+        contract_watcher: ContractStateWatcher,
+        mesh_state: watch::Receiver<MeshState>,
+        node_client: NodeClient,
+    ) -> anyhow::Result<Self> {
+        let Some(eth) = eth else {
+            tracing::warn!("ethereum indexer is disabled");
+            return Err(anyhow::anyhow!("ethereum indexer is disabled"));
+        };
+
+        let client = EthereumClient::new(eth.clone()).await?;
+
+        Ok(Self {
+            eth,
+            sign_tx,
+            app_data_storage,
+            node_near_account_id,
+            backlog,
+            contract_watcher,
+            mesh_state,
+            node_client,
+            client,
+        })
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub async fn run(&self) {
+        let backlog = self.backlog.clone();
+        let contract_watcher = self.contract_watcher.clone();
+        let mesh_state = self.mesh_state.clone();
+        let node_client = self.node_client.clone();
+        let sign_tx = self.sign_tx.clone();
+        let app_data_storage = self.app_data_storage.clone();
+        let client = self.client.clone();
```

Comment:
- too_many_arguments is not needed
- run should be owning self other we end up with these clones


```suggestion
    pub async fn run(self) {
```